keps/sig-storage/20200319-selinux-relabeling.md

```diff
@@ -0,0 +1,359 @@
+This KEP should follow API provided for fsGroup closely, however, the implementation is different (mount vs. recursive `chown`).
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### `mount -o context`
+Linux kernel, with SELinux compiled in, allows `mount -o context=s0:c309,c383 <what> <where>` to mount a volume and pretend that all files on the volume have given SELinux label.
+It works only for the first mount of the volume!
+It does not work for bind-mounts or any subsequent mount of the same volume.
+
+Note that volumes mounted with `-o context` don't have `seclabel` in their mount options.
+In addition, calling `chcon` there will fail with `Operation not supported`.
+
+### New Kubernetes behavior
+
+* When kubelet *knows* SELinux context of a pod / container to run (i.e. Pod/Container contains `SELinuxOptions`) and `SELinuxRelabelPolicy` is `OnVolumeMount`, it tries to mount all volumes for the Pod with given SELinux label using `mount -o context=XYZ`.
+    Kubelet makes sure that the option is passed to the first mount in all in-tree volume plugins (incl. ephemeral volumes like Secrets).
+    Kubelet passes it as an mount option to all CSI calls for given volume.
+
+    After the volume is mounted, kubelet checks that the root of the volume has the expected SELinux label, i.e. that the volume was mounted correctly.
+    * If the volume root has expected label, kubelet passes the volume to the container runtime without any ":z" or ":Z" options - no relabeling is necessary.
+    * If the volume root has unexpected label, for example when CSI driver did not apply `-o context` correctly or the volume was already mounted:
+        * If the volume supports SELinux (i.e. has `selabel` mount option in `/proc/mounts`), it passes ":Z" to the container runtime for the volume.
+            This is current kubelet behavior.
+        * If the volume does not support SELinux, it does not pass any ":Z" option to the container runtime.
+
+* Nothing changes if kubelet does not know the SELinux context of a pod (`SELinuxOptions` are empty),  kubelet passes ":Z" to the container runtime as today and lets the container runtime to choose a random label + relabel the volumes. 
+* Nothing changes if pod's `SELinuxRelabelPolicy` is `Always`, kubelet passes ":Z" to the container runtime as today.
+* Kubernetes validation check that `SELinuxRelabelPolicy` field can be used in a pod only when `SELinuxOptions` is set and SELinux label is known.
+
+#### Open Questions
+    When ... Pod/Container contains `SELinuxOptions`, ... [kubeket] tries to mount all volumes ... using `mount -o context=XYZ`.
+
+To mount with `-o context`, kubelet must know *all* `SELinuxOptions`, i.e. `User`, `Role`, `Type`, `Level`.
+Currently, when one or more of these fields are empty, the container runtime uses a default value set by OS.
+We need a way for kubelet to get these defaults.
```

Comment:
The SELinux go bindings can return you the content.  BTW Currently the only container label we use is

system_u:object_r:container_file_t:MCSLABEL

These labels are available in /etc/selinux/targeted/contexts/lxc_contexts, newer systems container-selinux packages now ship these labels in /usr/share/containers/selinux/context

Reply:
Great suggestion, added link to go-selinux and explained how the various label components are defaulted. In short, Level is never defaulted and must be non-empty in a Pod to trigger the new behavior. All other components can be empty.